enterprise/server/hit_tracker_service/hit_tracker_service.go

```diff
@@ -3,18 +3,49 @@ package hit_tracker_service
 import (
 	"context"
 
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
 	"github.com/buildbuddy-io/buildbuddy/server/real_environment"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
 
 	hitpb "github.com/buildbuddy-io/buildbuddy/proto/hit_tracker"
 )
 
 type HitTrackerService struct {
+	hitTrackerFactory interfaces.HitTrackerFactory
 }
 
-func Register(env *real_environment.RealEnv) {
-	env.SetHitTrackerServiceServer(&HitTrackerService{})
+func Register(env *real_environment.RealEnv) error {
+	if env.GetHitTrackerFactory() == nil {
+		return status.InvalidArgumentError("HitTrackerService requires a HitTrackerFactory")
+	}
+	env.SetHitTrackerServiceServer(&HitTrackerService{
+		hitTrackerFactory: env.GetHitTrackerFactory(),
+	})
+	return nil
 }
 
+// TODO(iain): record a source (e.g. Cache Proxy).
 func (h HitTrackerService) Track(ctx context.Context, req *hitpb.TrackRequest) (*hitpb.TrackResponse, error) {
+	for _, cacheHit := range req.GetHits() {
+		if cacheHit.GetInvocationId() == "" {
+			log.Debug("Skipping TrackRequest.Hits with empty invocation ID")
```

Comment:
nit: maybe warn if this indicates a bad request?

Reply:
Done